File: packages/shared/src/slog.ts
/**
 * Safe Structured Logging (ADR-12, SECURITY_PRIVACY.md §20, PI-06)
 * Accepts strictly whitelisted fields only to prevent accidental PII leakage.
 */

export const ALLOWED_LOG_FIELDS = new Set([
  'timestamp',
  'level',
  'module',
  'event',
  'session_id',
  'step_number',
  'action_type',
  'risk_category',
  'error_code',
  'duration_ms',
  'element_id',
  'status',
  'target_url_origin',
  'model_name',
  'sanitized_count',
  'success',
  'reason',
]);

export type LogLevel = 'debug' | 'info' | 'warn' | 'error';

export interface SafeLogEntry {
  level: LogLevel;
  module: string;
  event: string;
  session_id?: string;
  step_number?: number;
  action_type?: string;
  risk_category?: string;
  error_code?: string;
  duration_ms?: number;
  element_id?: string;
  status?: string;
  target_url_origin?: string;
  model_name?: string;
  sanitized_count?: number;
  success?: boolean;
  reason?: string;
  [key: string]: unknown;
}

export type LogSink = (entry: Record<string, unknown>) => void;

let currentSink: LogSink = (entry) => {
  const line = JSON.stringify(entry);
  if (entry.level === 'error') {
    process.stderr?.write ? process.stderr.write(line + '\n') : console.error(line);
  } else {
    process.stdout?.write ? process.stdout.write(line + '\n') : console.log(line);
  }
};

export function setLogSink(sink: LogSink): void {
  currentSink = sink;
}

export function filterSafeFields(entry: Record<string, unknown>): Record<string, unknown> {
  const filtered: Record<string, unknown> = {
    timestamp: typeof entry.timestamp === 'string' ? entry.timestamp : new Date().toISOString(),
  };

  for (const [key, value] of Object.entries(entry)) {
    if (ALLOWED_LOG_FIELDS.has(key)) {
      filtered[key] = value;
    }
  }

  return filtered;
}

export const slog = {
  log(level: LogLevel, entry: Omit<SafeLogEntry, 'level'>): Record<string, unknown> {
    const raw: Record<string, unknown> = { ...entry, level };
    const safe = filterSafeFields(raw);
    currentSink(safe);
    return safe;
  },

  info(entry: Omit<SafeLogEntry, 'level'>): Record<string, unknown> {
    return slog.log('info', entry);
  },

  warn(entry: Omit<SafeLogEntry, 'level'>): Record<string, unknown> {
    return slog.log('warn', entry);
  },

  error(entry: Omit<SafeLogEntry, 'level'>): Record<string, unknown> {
    return slog.log('error', entry);
  },

  debug(entry: Omit<SafeLogEntry, 'level'>): Record<string, unknown> {
    return slog.log('debug', entry);
  },
};

File: server/aegis_server/slog.py
"""
Safe Structured Logging for AEGIS Server (ADR-12, SECURITY_PRIVACY.md §20, PI-06)
Accepts strictly whitelisted fields only to prevent accidental PII leakage.
"""

from datetime import datetime, timezone
import json
import sys
from typing import Any, Callable, Dict, Optional

ALLOWED_LOG_FIELDS = {
    "timestamp",
    "level",
    "module",
    "event",
    "session_id",
    "step_number",
    "action_type",
    "risk_category",
    "error_code",
    "duration_ms",
    "element_id",
    "status",
    "target_url_origin",
    "model_name",
    "sanitized_count",
    "success",
    "reason",
}

LogSink = Callable[[Dict[str, Any]], None]


def _default_sink(entry: Dict[str, Any]) -> None:
    line = json.dumps(entry)
    if entry.get("level") == "error":
        sys.stderr.write(line + "\n")
        sys.stderr.flush()
    else:
        sys.stdout.write(line + "\n")
        sys.stdout.flush()


_current_sink: LogSink = _default_sink


def set_log_sink(sink: LogSink) -> None:
    global _current_sink
    _current_sink = sink


def filter_safe_fields(entry: Dict[str, Any]) -> Dict[str, Any]:
    safe: Dict[str, Any] = {
        "timestamp": entry.get("timestamp") or datetime.now(timezone.utc).isoformat(),
    }
    for k, v in entry.items():
        if k in ALLOWED_LOG_FIELDS:
            safe[k] = v
    return safe


class SafeLogger:
    @staticmethod
    def log(level: str, module: str, event: str, **kwargs: Any) -> Dict[str, Any]:
        raw = {"level": level, "module": module, "event": event, **kwargs}
        safe = filter_safe_fields(raw)
        _current_sink(safe)
        return safe

    @classmethod
    def info(cls, module: str, event: str, **kwargs: Any) -> Dict[str, Any]:
        return cls.log("info", module, event, **kwargs)

    @classmethod
    def warn(cls, module: str, event: str, **kwargs: Any) -> Dict[str, Any]:
        return cls.log("warn", module, event, **kwargs)

    @classmethod
    def error(cls, module: str, event: str, **kwargs: Any) -> Dict[str, Any]:
        return cls.log("error", module, event, **kwargs)

    @classmethod
    def debug(cls, module: str, event: str, **kwargs: Any) -> Dict[str, Any]:
        return cls.log("debug", module, event, **kwargs)


slog = SafeLogger()
